34-cellular-modbus-client/README.md

```diff
@@ -0,0 +1,427 @@
+- `seq_num`: A sequence number for the request. This number will be included in the response note in `responses.qo` so that requests can be matched to responses. The user should increment this value with every new request.
+- `func`: The Modbus function code.
+- `data`: A JSON object containing the function-specific data for the request. This object always contains an `addr` field, which specifies the start address of the coil(s)/register(s) to read/write.
+
+#### Request Types
+
+##### Read Coils
+
+```json
+{
+    "server_addr": 1,
+    "seq_num": 0,
+    "func": 1,
+    "data": {
+        "addr": 0,
+        "num_bits": 16
+    }
+}
+```
+
+`num_bits` is the number of coils to read, starting from `addr`.
+
+##### Read Discrete Inputs
+
+```json
+{
+    "server_addr": 1,
+    "seq_num": 1,
+    "func": 2,
+    "data": {
+        "addr": 0,
+        "num_bits": 16
+    }
+}
+```
+
+This is identical to the request for read coils, except the function code is 2 instead of 1.
+
+##### Read Holding Registers
+
+```json
+{
+    "server_addr": 1,
+    "seq_num": 3,
+    "func": 3,
+    "data": {
+        "addr": 0,
+        "num_regs": 2
+    }
+}
+```
+
+`num_regs` is the number of 16-bit holding registers to read, starting from `addr`.
+
+##### Read Input Registers
+
+```json
+{
+    "server_addr": 1,
+    "seq_num": 4,
+    "func": 4,
+    "data": {
+        "addr": 0,
+        "num_regs": 2
+    }
+}
+```
+
+This is identical to the request for read holding registers, except the function code is 4 instead of 3.
+
+##### Write Single Coil
+
+```json
+{
+    "server_addr": 1,
+    "seq_num": 5,
+    "func": 5,
+    "data": {
+        "addr": 2,
+        "val": 1,
+    }
+}
+```
+
+`addr` is the address of the coil to write and `val` is the value to write. Since coils are binary, `val` must be 0 or 1.
+
+##### Write Single Register
+
+```json
+{
+    "server_addr": 1,
+    "seq_num": 6,
+    "func": 6,
+    "data": {
+        "addr": 4,
+        "val": 199
+    }
+}
+```
+
+`addr` is the address of the register to write and `val` is the value to write. Since registers are 16-bits, `val` must fit into 16 bits. Note that only holding registers are writable. Input registers are read only.
+
+##### Write Multiple Coils
+
+```json
+{
+    "server_addr": 1,
+    "seq_num": 7,
+    "func": 15,
+    "data": {
+        "addr": 0,
+        "num_bits": 16,
+        "coil_bytes": [99, 171]
+    }
+}
+```
+
+`addr` is the start address of the coils to write. `num_bits` is the number of coils to write. `coil_bytes` is an array of bytes containing the values to write. The LSB of the first byte corresponds to the value to write at `addr`, the next most significant bit corresponds to `addr + 1`, and so on. The next byte's LSB corresponds to `addr + 8`, the next most significant bit of that byte to `addr + 9`, and so on. This table shows what values would be written to which addresses using the above example (in binary, 99 is 01100011 and 171 is 10101011):
```

Comment:
> Loose opinion - it kind of feels backwards, in that, it's more intuitive (at least to me) that the MSB is the first bit written.

Yeah, I think it's just a matter of opinion. My take is that we call the LSB bit 0, so that should correspond to address 0, the next bit to address 1, and so on. Prefer to leave as-is, unless you feel strongly about changing it.

Reply:
No strong opinion here, happy to leave as is, especially if that's the precedence in the Modbus space.